src/mapping_navigation/scripts/pure_pursuit.py

```diff
@@ -0,0 +1,152 @@
+        Guillaume Jacquenot (@Gjacquenot)
+"""
+import numpy as np
+import math
+from typing import List, Tuple
+from map_information import *
+
+class State:
+
+    """
+    Contains the car information needed in pure pursuit
+    The cartesian coordinates for location
+    The yaw of the car
+    The cars wheel base
+    """
+    def __init__(self, wheel_base: float, x: float=0.0, y:float=0.0, yaw:float=0.0, velocity:float=0.0):
+        self.wheel_base = wheel_base
+        self.x = x
+        self.y = y
+        self.yaw = yaw
+        self.velocity = velocity
+        self.rear_x = self.x - ((wheel_base / 2) * math.cos(self.yaw))
+        self.rear_y = self.y - ((wheel_base / 2) * math.sin(self.yaw))
+    
+    """
+    Updates the cars position
+    Cartesian coordinates for the cars position
+    Quaterion to find the yaw of the car
+    """
+    def update(self, x: float, y: float, quaterion: Tuple[float, float, float, float], velocity:float):
+        self.x = x
+        self.y = y
+        self.velocity = velocity
+        self.yaw = euler_from_quaternion(quaterion)
+        self.rear_x = self.x - ((self.wheel_base / 2) * math.cos(self.yaw))
+        self.rear_y = self.y - ((self.wheel_base / 2) * math.sin(self.yaw))
+
+    """
+    Find the distance between the cars current location and another point
+    Find the euclidian distance
+    """
+    def calc_distance(self, point_x:float, point_y:float) -> float:
+        dx = self.rear_x - point_x
+        dy = self.rear_y - point_y
+        return math.hypot(dx, dy)
+
+class PurePursuit:
+
+    """
+    look aheaed distiance is the distance pur pursuit scans to find goals points
+    look forward gain is used to increase the look ahead distance based on the cars current speed
+    path is the path the algorithm must track
+    """
+    def __init__(self, look_ahead_distance: float, look_forward_gain: float, path :List[Point]):
+        self.look_forward_gain = look_forward_gain
+        self.look_ahead_distance = look_ahead_distance
+        self.path = path
+        #Needs to be refactored to look at our points instead of his cx and cy
+        self.old_nearest_point_index = None
+        self.old_steer_target = 0
+        self.cx = []
+        self.cy = []
+        for point in path:
+            self.cx.append(point.coordinate[0])
+            self.cy.append(point.coordinate[1])
+
+    """
+    Returns the steering angle in radians and the index to the goal points its turning to
+    Takes in the current state of the car
+    """
+    def pure_pursuit_steer_control(self, state: State) -> Tuple[float, int]:
+        
+        # Getting the next point
+        ind, Lf = self.search_target_index(state)
+
+        # Checks to see if the previous found goal points is actually ahead of what it sees now
+        # If the previous index is ahead it uses that over the new one.
+        if self.old_steer_target >= ind:
+            ind = self.old_steer_target
+
+        if ind < len(self.cx):
```

Comment:
What is this branch representing? Other is towards goal, so not sure what this is for

Reply:
this should be removed, its a hangover from the way it was originally implemented that I forgot to get rid of, the guy has it running in a loop, so he is just resetting it here, my bad forgot about that.